File: home/gui/eww/eww/scripts/weather.py
import sys
import requests
from subprocess import run

API_KEY = "aba36eff207740e0a5e234935232607"
CITY_NAME = "Silistra"
color = "#ffffff"

url = f"https://api.weatherapi.com/v1/current.json?key={API_KEY}&q={CITY_NAME}"
response = requests.get(url)
data = response.json()

temperature = data["current"]["temp_c"]
status = data["current"]["condition"]["text"]
icon = data["current"]["condition"]["code"]
wind_speed = data["current"]["wind_kph"]
is_day = data["current"]["is_day"]

if icon == 1000: icon = ""; color = "#ffb78a"
elif icon == 1003: icon = ""; color = "#b1e4eb"
elif icon == 1006: icon = ""; color = "#b1e4eb"
elif icon == 1009: icon = ""; color = "#b1e4eb"
elif icon == 1030: icon = ""; color = "#c9c9c9"
elif icon == 1063: icon = ""; color = "#83ace2"
elif icon == 1066: icon = ""; color = "#cad2db"
elif icon == 1069: icon = ""; color = "#83ace2"
elif icon == 1072: icon = ""; color = "#83ace2"
elif icon == 1087: icon = ""; color = "#eff0b3"
elif icon == 1114: icon = ""; color = "#cad2db"
elif icon == 1117: icon = ""; color = "#cad2db"
elif icon == 1135: icon = ""; color = "#c9c9c9"
elif icon == 1147: icon = ""; color = "#c9c9c9"
elif icon == 1150: icon = ""; color = "#83ace2"
elif icon == 1168: icon = ""; color = "#83ace2"
elif icon == 1171: icon = ""; color = "#83ace2"
elif icon == 1180: icon = ""; color = "#83ace2"
elif icon == 1183: icon = ""; color = "#83ace2"
elif icon == 1189: icon = ""; color = "#83ace2"
elif icon == 1192: icon = ""; color = "#83ace2"
elif icon == 1195: icon = ""; color = "#83ace2"
elif icon == 1198: icon = ""; color = "#83ace2"
elif icon == 1201: icon = ""; color = "#83ace2"
elif icon == 1204: icon = ""; color = "#83ace2"
elif icon == 1207: icon = ""; color = "#83ace2"
elif icon == 1210: icon = ""; color = "#b1e4eb"
elif icon == 1213: icon = ""; color = "#cad2db"
elif icon == 1216: icon = ""; color = "#b1e4eb"
elif icon == 1219: icon = ""; color = "#cad2db"
elif icon == 1222: icon = ""; color = "#cad2db"
elif icon == 1225: icon = ""; color = "#cad2db"
elif icon == 1237: icon = ""; color = "#cad2db"
elif icon == 1240: icon = ""; color = "#83ace2"
elif icon == 1243: icon = ""; color = "#83ace2"
elif icon == 1246: icon = ""; color = "#83ace2"
elif icon == 1249: icon = ""; color = "#83ace2"
elif icon == 1252: icon = ""; color = "#83ace2"
elif icon == 1255: icon = ""; color = "#cad2db"
elif icon == 1258: icon = ""; color = "#cad2db"
elif icon == 1261: icon = ""; color = "#cad2db"
elif icon == 1264: icon = ""; color = "#cad2db"
elif icon == 1273: icon = ""; color = "#eff0b3"
elif icon == 1276: icon = ""; color = "#eff0b3"
elif icon == 1279: icon = ""; color = "#eff0b3"
elif icon == 1282: icon = ""; color = "#eff0b3"

if is_day == 0:
    if icon == "": icon = ""; color = "#ffb78a"
    elif icon == "": icon = ""; color = "#b1e4eb"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#b1e4eb"
    elif icon == "": icon = ""; color = "#b1e4eb"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"
    elif icon == "": icon = ""; color = "#83ace2"

if sys.argv[1] == "temp": print(f"{temperature} °C")
elif sys.argv[1] == "status": print(f"{status}")
elif sys.argv[1] == "wind": print(f"Wind: {wind_speed}km/h")
elif sys.argv[1] == "icon": print(icon)
elif sys.argv[1] == "icon-color": print (color)
elif sys.argv[1] == "day": print (is_day)

File: home/gui/eww/eww/scripts/weather.c
#!/usr/bin/env scriptisto

// scriptisto-begin
// script_src: weather.c
// build_cmd: gcc weather.c -o ./script
// scriptisto-end

#include <stdio.h> 
#include <stdlib.h>
#include <string.h>

char* get_attr(char* set, char* attr) {
    char* start = strstr(set, attr);
    start += strlen(attr) + 2;

    int is_str = 0;
    if (*start == '"') {
        start++;
        is_str = 1;
    }

    char* p = start;
    size_t len = 0;

    while (*p != ',' && *p != '}') {
        len++;
        p++;
    }

    if (is_str) {
        len--;
    }

    char* attr_val = malloc(len + 1);
    strncpy(attr_val, start, len);
    attr_val[len] = '\0';

    return attr_val;
}
  
int main() 
{
    char* key = "aba36eff207740e0a5e234935232607";
    char* city = "Silistra";
    char* cmd_fmt = "curl -s \"https://api.weatherapi.com/v1/current.json?key=%s&q=%s\"";
    size_t cmd_len = strlen(cmd_fmt) + strlen(key) + strlen(city);
    char* cmd = malloc(cmd_len+1);
    sprintf(cmd, cmd_fmt, key, city);

    FILE* p = popen(cmd, "r");

    char* resp = NULL;
    size_t resp_size = 0;
    if (getline(&resp, &resp_size, p) != -1) {
        char* temp = get_attr(resp, "temp_c");
        char* status = get_attr(resp, "text");
        char* icon_code = get_attr(resp, "code");
        char* wind_speed = get_attr(resp, "wind_kph");
        char* is_day = get_attr(resp, "is_day");

        char* icon_color = NULL;
        char* icon = NULL;

        if (!strcmp(icon_code, "1000")) { icon = ""; icon_color = "#ffb78a"; }
        else if (!strcmp(icon_code, "1003")) { icon = ""; icon_color = "#b1e4eb"; }
        else if (!strcmp(icon_code, "1006")) { icon = ""; icon_color = "#b1e4eb"; }
        else if (!strcmp(icon_code, "1009")) { icon = ""; icon_color = "#b1e4eb"; }
        else if (!strcmp(icon_code, "1030")) { icon = ""; icon_color = "#c9c9c9"; }
        else if (!strcmp(icon_code, "1063")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1066")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1069")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1072")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1087")) { icon = ""; icon_color = "#eff0b3"; }
        else if (!strcmp(icon_code, "1114")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1117")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1135")) { icon = ""; icon_color = "#c9c9c9"; }
        else if (!strcmp(icon_code, "1147")) { icon = ""; icon_color = "#c9c9c9"; }
        else if (!strcmp(icon_code, "1150")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1168")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1171")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1180")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1183")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1189")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1192")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1195")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1198")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1201")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1204")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1207")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1210")) { icon = ""; icon_color = "#b1e4eb"; }
        else if (!strcmp(icon_code, "1213")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1216")) { icon = ""; icon_color = "#b1e4eb"; }
        else if (!strcmp(icon_code, "1219")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1222")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1225")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1237")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1240")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1243")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1246")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1249")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1252")) { icon = ""; icon_color = "#83ace2"; }
        else if (!strcmp(icon_code, "1255")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1258")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1261")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1264")) { icon = ""; icon_color = "#cad2db"; }
        else if (!strcmp(icon_code, "1273")) { icon = ""; icon_color = "#eff0b3"; }
        else if (!strcmp(icon_code, "1276")) { icon = ""; icon_color = "#eff0b3"; }
        else if (!strcmp(icon_code, "1279")) { icon = ""; icon_color = "#eff0b3"; }
        else if (!strcmp(icon_code, "1282")) { icon = ""; icon_color = "#eff0b3"; }
        else { icon = ""; icon_color = "#83ace2"; }

        if (!strcmp(is_day, "0")) {
            if (!strcmp(icon, "")) { icon = ""; icon_color = "#83ace2"; }
            else { icon = ""; icon_color = "#b1e4eb"; }
        }

        printf
            (
                "{\"temp\": \"%s\", \"status\": \"%s\", \"icon\": \"%s\", \"icon_color\": \"%s\", \"wind_speed\": \"%s\"}\n",
                    temp,             status,             icon,             icon_color,             wind_speed
            );

        free(temp);
        free(status);
        free(icon_code);
        free(wind_speed);
        free(is_day);
    }

    free(resp);
    free(cmd);
    pclose(p);
    return 0;
}

